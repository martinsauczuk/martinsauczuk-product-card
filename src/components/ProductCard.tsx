import React, { CSSProperties, createContext } from 'react';
import { useProduct } from '../hooks/useProduct';
import { Product, ProductContextProps, OnChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';
import styles from '../styles/styles.module.css';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;


export interface Props {
    // children?: ReactElement | ReactElement[];
    children: ( args: ProductCardHandlers ) => JSX.Element;
    product: Product;
    className?: string;
    style?: CSSProperties;
    onChange?: ( args: OnChangeArgs ) => void;
    value?: number;
    initialValues?: InitialValues
}

export const ProductCard = ({children, product, className, style, onChange, value, initialValues }: Props ) => {

    const { increaseBy, counter, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });

    return (

        <Provider value = {{
            counter,
            maxCount,
            increaseBy,
            product,
        }}>
            <div 
                className={ `${styles.productCard} ${className}` }
                style={style}
            >
                { 
                    children({
                        count: counter,
                        isMaxCountReached,
                        maxCount: initialValues?.maxCount,
                        product,
                        increaseBy,
                        reset,
                    })
                }
            </div>
        </Provider>
    )
}
