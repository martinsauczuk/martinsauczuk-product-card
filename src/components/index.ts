export { ProductButtons } from "./ProductButtons";
import { ProductImage } from ".";
import { ProductButtons } from ".";
import { ProductTitle } from ".";
import { ProductCardHOCProps } from "../interfaces/interfaces";
import { ProductCard as ProductCardHOC } from './ProductCard';
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard: ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
});

export default ProductCard;