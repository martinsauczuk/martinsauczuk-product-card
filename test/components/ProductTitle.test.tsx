import React from "react";
import renderer from "react-test-renderer";
import { ProductTitle, ProductCard } from '../../src/components';
import { product1 } from '../data/products';

describe('ProductTitle', () => {

  test('should show the component with custom title', () => {
    const wrapper = renderer.create(
        <ProductTitle title="Custom Product"/>
    )
    expect( wrapper.toJSON() ).toMatchSnapshot();
  });

  test('should show the componente with the product name', () => {
    
    const wrapper = renderer.create(
        <ProductCard product={ product1 } >
            {
                () => <ProductTitle />
            }
        </ProductCard>
    )
    expect( wrapper.toJSON() ).toMatchSnapshot();

  });
  
})
