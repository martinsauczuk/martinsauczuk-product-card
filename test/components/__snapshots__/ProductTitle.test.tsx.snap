// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle should show the component with custom title 1`] = `
<span
  className="productDescription undefined"
>
  Custom Product
</span>
`;

exports[`ProductTitle should show the componente with the product name 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Cofree Mug / Card without image
  </span>
</div>
`;
